jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/lock/DebugRuntimeManagerLock.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.runtime.manager.impl.lock;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.jbpm.runtime.manager.spi.RuntimeManagerLock;
+
+
+public class DebugRuntimeManagerLock implements RuntimeManagerLock {
+
+    private ReentrantLock lock = new ReentrantLock(true);
+    private List<StackTraceElement[]> traces;
+    private long currentThreadId;
+
+    public DebugRuntimeManagerLock() {
+        traces = new ArrayList<>();
+        currentThreadId = -1;
+    }
+    
+    @Override
+    public void lock() {
+        lock.lock();
+        addTraces();
+    }
+
+    @Override
+    public boolean tryLock(long units, TimeUnit timeUnit) throws InterruptedException {
+        boolean outcome = lock.tryLock(units, timeUnit);
+        if(outcome) {
+            addTraces();
+        }
+        return outcome;
+    }
+
+    @Override
+    public void lockInterruptible() throws InterruptedException {
+        lock.lockInterruptibly();
+        addTraces();
+    }
+
+    private void addTraces() {
+        long newThreadId = Thread.currentThread().getId();
+        if(currentThreadId < 0 || currentThreadId != newThreadId) {
+            currentThreadId = newThreadId;
+            traces.clear();
+        }
+        // we add all stack traces during reentrant lock
+        traces.add(Thread.currentThread().getStackTrace());
+    }
+    @Override
+    public void unlock() {
+        if(currentThreadId == Thread.currentThread().getId()) {
+            traces.clear(); // clean up
+        }
+        lock.unlock();
```

Comment:
if there is a place where lock for the same process instance id in the same thread is calling more than once. the logic of that part is broken. A thread (request execution) should only invoke once the lock. That is also the reason a added more that one stack traces... to add diaganose that possibility (more than one call to lock in one request)

That is pretty good idea actually

Reply:
done

Reply:
Could we use here just `getHoldCount()` and clear traces only if it is equals to 1? Maybe we wouldn't need an additional counter for that. As right now when we have 2 locks called from the same thread (I know it shouldn't happen), the counter is decremented in a non-thread-safe way I think. Same goes for increments in case there are 2 lock operations called by the same thread. Also the counter should be probably reset in case a new thread has acquired the lock. I think getHoldCount() would be OK for this as it uses the internal counter of the lock.

Reply:
will change